src/tests/system/tests/test_gdm.py

```diff
@@ -0,0 +1,114 @@
+    """
+    :title: Login via GDM with external IdP user is rejected with invalid password
+    :setup:
+        1. Configure IPA for External IdP support
+        2. Add user to IPA as IdP user with authtype only set to 'idp'
+        3. Configure SSSD pam_json_services = gdm-switchable-auth
+    :steps:
+        1. Select user from list (if not listed, use Other and enter user)
+        2. Follow IdP redirect url and use invalid password to login via External IdP site
+        3. Check logs for login denial
+    :expectedresults:
+        1. User selected and login method (EIdP) chosen
+        2. User should see login rejected/denied error for invalid password
+        3. Should see login denial in logs
+    """
+    testuser = "kcgdmuser1"
+    password = "Secret123"
+    testuser_idp = f"{testuser}@{keycloak.host.hostname}"
+
+    client.sssd.import_domain('gdm.test', ipa)
+    client.sssd.pam["pam_json_services"] = "gdm-switchable-auth"
+    client.sssd.start()
+
+    keycloak.user(testuser).add()
+    keycloak.user(testuser).modify(email=f"{testuser_idp}")
+    ipa.user(testuser).add()
+    ipa.user(testuser).modify(user_auth_type="idp", idp="keycloak", idp_user_id=testuser_idp)
+
+    assert not client.gdm.login_idp(client, testuser, password[:-1])
+
+
+@pytest.mark.topology(KnownTopology.GDM)
+def test_gdm__xidp_login_get_kerberos_ticket(client: Client, ipa: IPA, keycloak: Keycloak):
+    """
+    :title: Login via GDM with external IdP user
+    :setup:
+        1. Configure IPA for External IdP support
+        2. Add user to IPA as IdP user with authtype only set to 'idp'
+        3. Configure SSSD pam_json_services = gdm-switchable-auth
+    :steps:
+        1. Select user from list (if not listed, use Other and enter user)
+        2. Follow IdP redirect url and login as user via External IdP site
+        3. Complete login in GDM (select Done)
+        4. Confirm user logged in by checking for home screen
+        5. Open terminal and check for Kerberos ticket
+    :expectedresults:
+        1. User selected and login method (EIdP) chosen
+        2. User connects to remote URL with browser and enters login info successfully
+        3. User is logged into system
+        4. User can see login home screen after login
+        5. User can see Kerberos ticket has been issued
+    """
+    testuser = "kcgdmuser1"
+    password = "Secret123"
+    testuser_idp = f"{testuser}@{keycloak.host.hostname}"
+
+    client.sssd.import_domain('gdm.test', ipa)
+    client.sssd.pam["pam_json_services"] = "gdm-switchable-auth"
+    client.sssd.start()
+
+    keycloak.user(testuser).add()
+    keycloak.user(testuser).modify(email=f"{testuser_idp}")
+    ipa.user(testuser).add()
+    ipa.user(testuser).modify(user_auth_type="idp", idp="keycloak", idp_user_id=testuser_idp)
+
+    assert client.gdm.login_idp(client, testuser, password, close=False)
+
+    klist = client.host.conn.run(f'su - {testuser} -c klist')
```

Comment:
KerberosAuthenticationUtils seems to provide some ways to check whether a user has been provided a Kerberos ticket. I'd recommend you using one of them, and if none serves your purpose you can implement one.

Reply:
still working on this one.   I hit some issues with test environment and didn't finish this one yet.   Will hopefully have an update tomorrow.

Reply:
added a "has_tgt" function that is now used here.